### DISCUSSION TITLE
My application using too much memory due to background tasks (BackgroundTask). How can I monitor and control them to prevent a crash.

### DISCUSSION THREAD

**Discussion by ghafooretemad** [fastapi/fastapi#11156] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, BackgroundTasks
import time

app = FastAPI()

def run_heavy_task(data):
    print(data)
    time.sleep(10)
    
@app.post("/heavy-tasks")
def do_some_heavy_task(data:str, background_task: BackgroundTasks=BackgroundTasks):
    background_task.add_task(run_heavy_task, data)
    return {"msg": "data sent to the server"}
```

### Description

My application receives continuous data streams from various nodes, each requiring database connections, data insertion and queries. These operations are handled by background tasks. Can I implement a mechanism to control the number of waiting tasks to prevent memory overload?

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.10.12

**Comment by jgould22** [comment, marked as answer]
Personally I think background tasks are nice to have for lightweight things like book keeping after a request etc.

The second I am doing more heavy weight stuff you should probably reach for Celery and the like even if it add considerable complication to a project.

**Reply by ghafooretemad** [reply]
Agreed. However, having a mechanism for accessing the number of running background tasks allows us to overcome this issue without adding complexity to the project (like Celery).
